src/dataPipeline.js

```diff
@@ -5,166 +5,520 @@
+ * @return {PenultimateNavigation} The results of the navigation, including members `pen`, `root`, `last`
+ */
+fluid.data.getPenultimate = function (root, segs) {
+    var move = root,
+        prev, seg;
+    for (var i = 0; i < segs.length; ++i) {
+        prev = move;
+        seg = segs[i];
+        move = move && move[seg];
+    }
+    return {
+        pen: prev,
+        root: move,
+        last: seg
+    };
+};
+
+/** Do the work of STRUCTURAL PROMOTION for "short" records that are found to the right of
+ * of a taller one. So far we have discovered a `fluid.compoundElement` at path `path` and index `currentLayer` and we
+ * now scan elements to the right which are not compound in order to upgrade them.
+ * @param {Object[]} layers - The unmerged `layers` records found within the `fluid.compoundElement` parent's merging options.
+ * **A layer in this structure may be modified through the structural promotion operation **
+ * @param {Integer} currentLayer - The layer number at which we found a compound element scanning to the left
+ * @param {String[]} path - Array of path segments of the location at which the compound element was found
+ */
+fluid.data.upgradeElements = function (layers, currentLayer, path) {
+    for (var i = currentLayer + 1; i < layers.length; ++i) {
+        var thisLayer = layers[i];
+        var pen = fluid.data.getPenultimate(thisLayer, path);
+        var current = pen.root;
+        // Taller than the tallest tree, noone here's as tall as me
+        if (fluid.isPlainObject(current, true) && current.type !== "fluid.compoundElement") {
+            var upgraded = {
+                elements: {
+                    [pen.last]: current
+                }
+            };
+            pen.pen[pen.last] = upgraded;
+            console.log("After upgrade, layer " + i + " reads ", thisLayer);
+        }
+    }
+};
+
+/** Given an array of element definitions, resolve their fully merged structure by means of STRUCTURAL PROMOTION
+ * @param {Object[]} layers - Array of unmerged element definitions
+ * @return {Object} A merged hash of element definitions, also convered into standard Infusion grade definitions
+ */
+fluid.data.mergeElements = function (layers) {
+    // Elements on the left composite on top of elements to the right
+    for (var i = 0; i < layers.length; ++i) {
+        var layer = layers[i];
+        var path = [];
+        fluid.each(layer, function (record, key) { // eslint-disable-line no-loop-func
+            path.push(key);
+            // We've found a compound - check all elements to the left in reverse order to see if they need an upgrade
+            if (fluid.isPlainObject(record, true) && record.type === "fluid.compoundElement") {
+                fluid.data.upgradeElements(layers, i, path);
+            }
+            path.pop();
+        });
+    }
+    var mergedElements = fluid.extend(true, {}, ...layers);
+    console.log("Got mergedElements ", JSON.stringify(mergedElements, null, 2));
```

Comment:
Is this a debugging line? Do you prefer to keep it?

Reply:
There's a lot of debug output generally since this is still pretty immature code and it is good to know what is happening. It's particularly interesting to see the "mergedElements" since this is the only point that the fully merged pipeline can be seen (which would be the main viewable artefact in any GUI for this)